mesop/dataclass_utils/dataclass_utils.py

```diff
@@ -127,3 +139,73 @@ def decode_mesop_json_state_hook(dct):
   except ImportError:
     pass
   return dct
+
+
+class DataFrameOperator(BaseOperator):
+  """Custom operator to detect changes in DataFrames.
+
+  DeepDiff does not support diffing DataFrames. See https://github.com/seperman/deepdiff/issues/394.
+
+  This operator checks if the DataFrames are equal or not. It does not do a deep diff of
+  the contents of the DataFrame.
+  """
+
+  def match(self, level) -> bool:
+    try:
+      import pandas as pd
+
+      return isinstance(level.t1, pd.DataFrame) and isinstance(
+        level.t2, pd.DataFrame
+      )
+    except ImportError:
+      # If Pandas is not installed, don't perform this check. We should log a warning.
+      return False
+
+  def give_up_diffing(self, level, diff_instance) -> bool:
+    if not level.t1.equals(level.t2):
```

Comment:
Just double checking - is this using https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.equals.html#pandas-dataframe-equals which is doing a deep comparison check (and not an identity check)?

Reply:
Yeah, initially I was trying to to level.t1 == level.t2, but it kept failing with "ValueError: The truth value of a DataFrame is ambiguous".

I wasn't quite sure how to resolve it. Probably need to look into it more.

Reply:
ah ok, using .equals seems good to me.